feat: Add file watcher for pending deletions and 7-day log rotation

vamsi-op opened the pull request:
## Description

This PR implements the usability improvements requested in #5:

1. **File System Watcher for Pending Deletions**
   - Spawns a background daemon process that monitors all files/directories scheduled for deletion
   - Uses watchdog library for efficient cross-platform file system event monitoring
   - Automatically detects when pending files/directories are manually deleted
   - Cleans up pending.json entries when deletions are detected
   - No manual intervention required

2. **Automatic Log Rotation (7-day cleanup)**
   - Updated logger to use TimedRotatingFileHandler
   - Logs rotate every 7 days automatically
   - Keeps 4 weeks (28 days) of log history
   - Old logs are automatically removed

## Implementation Details

### New Files
- src/dswell/watcher.py: Core watcher implementation
- src/dswell/subcmds/watcher.py: CLI commands for watcher management
- tests/test_watcher.py: Comprehensive test suite (15 tests)

### Modified Files
- pyproject.toml: Added watchdog>=3.0.0 dependency
- src/dswell/logger.py: Changed to TimedRotatingFileHandler with 7-day rotation
- src/dswell/cli.py: Registered watcher subcommand
- README.md: Added documentation for new features

## Usage

\\\ash
# Start the file watcher daemon
dswell watcher start

# Check watcher status
dswell watcher status

# Stop the watcher daemon
dswell watcher stop
\\\

## Testing

- All existing tests pass
- 15 new tests added for watcher functionality
- Test coverage for file deletion events, PID management, daemon control

## Closes #5

## Summary by Sourcery

Implement a cross-platform file system watcher daemon for pending deletions and add automatic 7-day log rotation, register corresponding CLI commands, update dependencies, documentation, and tests.

New Features:
- Add a background file watcher daemon to monitor and clean up pending deletions
- Provide watcher CLI commands (start, stop, status) to manage the watcher process
- Introduce automatic log rotation every 7 days with 4 weeks of retention via TimedRotatingFileHandler

Enhancements:
- Register the watcher subcommand in the main CLI
- Use the watchdog library for efficient cross-platform file system event monitoring

Build:
- Add watchdog>=3.0.0 dependency

Documentation:
- Document the file watcher commands and automatic log cleanup in README

Tests:
- Add a comprehensive test suite (15 tests) for watcher functionality, including event handling, daemon control, and CLI commands

vamsi-op committed:
feat: add file watcher for pending deletions and 7-day log rotation

- Add watchdog dependency for file system monitoring
- Implement PendingWatcher daemon to monitor scheduled deletions
- Automatically clean pending.json when files are manually deleted
- Add watcher CLI commands (start/stop/status)
- Update logger to use TimedRotatingFileHandler with 7-day rotation
- Keep 4 weeks of log history (28 days total)
- Add comprehensive tests for watcher functionality (15 tests)
- Update README with watcher usage and features documentation

Closes #5

vamsi-op commented:
## Response to Sourcery Bot Review

Thank you for the detailed review! Here are my responses to each comment:

### 1. Question: Skipping non-existent parent directories

Good catch! The current implementation is intentional for the initial version. When a parent directory is deleted before the watcher starts, the pending entry becomes invalid anyway since the file no longer exists. The watcher's primary purpose is to detect **manual** deletions that happen **after** files are scheduled.

However, for completeness, I'll add logging when parent directories don't exist so users are aware. This is acceptable behavior since:
- If the parent dir is gone, the file is already gone
- The daemon process for that file will clean up the pending.json entry when it tries to delete and finds the file missing

### 2. Issue: Relying on internal 'emitters' attribute

Excellent point about fragility. You're right that accessing internal attributes like `emitters` and `watch` is risky. I'll refactor this to maintain our own set of watched paths internally. This will be more maintainable and won't break with watchdog library updates.

### 3. Suggestion: status_watcher() output

Actually, `status_watcher()` **does** print output! If you check the implementation in `src/dswell/watcher.py` lines 259-271, it uses `logger.info()` which outputs to console via the console handler. The messages are:
- "Watcher is running (PID: {pid})"
- "Watcher is not running"

The logger is configured with a console handler at INFO level, so these messages are visible to users. However, I can switch to `click.echo()` for more direct CLI output if preferred.

### 4. Suggestion: Missing test for error handling in _update_monitored_paths

Great suggestion! I'll add this test case to improve error handling coverage.

### 5. Suggestion: No test for deletion of a path not in monitored_paths

Good catch! This edge case should be tested. I'll add it.

### 6. Suggestion: No test for PID file write failure

Excellent point for robustness testing. I'll add this test case.

### 7. Suggestion: No test for is_running with invalid PID file contents

This is already handled in the code (`except (OSError, ValueError)` in `is_running()`), but you're right that we should have an explicit test for it. I'll add it.

### 8. Code Quality: Use any() instead of for loop

Perfect refactoring suggestion! This is more Pythonic and cleaner. I'll apply it.

### 9. Code Quality: Extract code into function

Good suggestion for the stop_watcher logic. I'll extract it into a helper function for better code organization.

---

I'll push these improvements shortly. Thanks for the thorough review!

Agent-Hellboy commented:
@vamsi-op, please refrain from using AI. You are contributing to open-source for your learning, not to increase PR counts.

vamsi-op committed:
refactor: address Sourcery bot review suggestions

- Add internal tracking of watched paths instead of accessing observer.emitters
- Add logging when parent directories don't exist
- Add test for deletion of non-existent path
- Add test for invalid PID file contents
- Improve code maintainability and test coverage

Addresses Sourcery bot review comments on PR #9

vamsi-op commented:
> > @vamsi-op, please refrain from using AI. You are contributing to open-source for your learning, not to increase PR counts.
  
    I appreciate your concern!
    I noticed Sourcery AI reviewed the PR, I actually implemented its suggestions to improve code quality! I'm using AI the same way :)

If that's the case, I could have used AI and added a PR it's for your learning, man You should verify everything manually, every piece of code, and if you have this much confidence in AI, you can try  https://github.com/Agent-Hellboy/gunicorn-prometheus-exporter/issues/44 and https://github.com/Agent-Hellboy/gunicorn-prometheus-exporter/issues/67 and see for yourself, no LLM model would solve this.

vamsi-op committed:
chore: update watchdog to latest version 6.0.0

Agent-Hellboy commented:
What do you understand about log rotation? Have you heard this term before? How could you simply throw whatever AI is throwing? Do you understand anything about watchdog? 
Please at least explore things for 1-2 days, try to understand it, man. You could have learned a lot if you had tried to code it using a conventional way.
